File: progcomp/3144.rs
use std::io;

#[derive(PartialOrd, PartialEq, Debug, Ord, Eq)]
struct Edge {
    w: i32,
    u: i32,
    v: i32,
}

fn find(u: i32, id: &mut Vec<i32>) -> i32 {
    if id[u as usize] == u {
        id[u as usize] = u;
    } else {
        id[u as usize] = find(id[u as usize], id);
    }
    id[u as usize]
}

fn uni(u: i32, v: i32, id: &mut Vec<i32>, sz: &mut Vec<i32>) -> Option<i32> {
    let mut a = find(u, id);
    let mut b = find(v, id);
    if a == b {
        return None;
    }
    if sz[a as usize] > sz[b as usize] {
        let aux = a;
        a = b;
        b = aux;
    }
    sz[a as usize] += sz[b as usize];
    id[b as usize] = a;
    Some(1)
}

fn main() {
    let mut input = String::new();
    let _ = io::stdin().read_line(&mut input);
    let mut iter = input.split_whitespace();
    let N: i32 = iter.next().unwrap().parse().unwrap();
    let M: i32 = iter.next().unwrap().parse().unwrap();
    let _ = io::stdin().read_line(&mut input);
    let mut edges: Vec<Edge> = Vec::new();
    let mut sz = vec![1; (N + 4) as usize];
    let mut id: Vec<i32> = (0..=(N + 4)).collect();
    for _ in 0..M {
        input = "".to_string();
        let _ = io::stdin().read_line(&mut input);
        iter = input.split_whitespace();
        let u: i32 = iter.next().unwrap().parse().unwrap();
        let v: i32 = iter.next().unwrap().parse().unwrap();
        let w: i32 = iter.next().unwrap().parse().unwrap();
        edges.push(Edge { w, u, v });
    }
    edges.sort();
    let mut ans = 0;
    for edg in edges {
        let u = edg.u;
        let v = edg.v;
        if find(u, &mut id) == find(v, &mut id) {
            continue;
        }
        uni(u, v, &mut id, &mut sz);
        ans += edg.w;
    }
    println!("{:?}", 2 * ans);
}

File: progcomp/3144.py
line = input()
n, m = line.split(" ")
n = int(n)
m = int(m)
line = input()
edges = []
for _ in range(m):
    line = input()
    u, v, w = line.split(" ")[0:3]
    u = int(u)
    v = int(v)
    w = int(w)
    edges.append((u, v, w))
edges.sort(key=lambda x: x[2])

id = [x for x in range(n + 1)]
sz = [1 for _ in range(n + 1)]


def find(u):
    if u == id[u]:
        return u
    id[u] = find(id[u])
    return id[u]


def uni(u, v):
    u = find(u)
    v = find(v)
    if u == v:
        return
    if sz[u] > sz[v]:
        u, v = v, u
    sz[u] += sz[v]
    id[v] = u


ans = 0
for edg in edges:
    if find(edg[0]) == find(edg[1]):
        continue
    ans += edg[2]
    uni(edg[0], edg[1])

print(ans * 2)
